### PR TITLE
First commit for event index

### PR DESCRIPTION
**Pull request by justin13601**
For #37 

## Summary by CodeRabbit

- **New Features**
  - Introduced event indexing with a new `_EVENT_INDEX` column in various aggregation and predicate functions.
  - Enhanced extraction and processing of subtree data with event indexing.

- **Improvements**
  - Enhanced data consistency by including event index columns in summary outputs.

- **Tests**
  - Updated end-to-end tests to include new `_LAST_EVENT_INDEX` fields.

### CODE REVIEW

File: src/aces/aggregate.py [hunk 1]
```diff
@@ -305,71 +315,74 @@ def aggregate_event_bound_window(
         ...     "is_A": [1, 0, 1, 1, 1, 1, 0, 0],
         ...     "is_B": [0, 1, 0, 1, 0, 1, 1, 1],
         ...     "is_C": [0, 1, 0, 0, 0, 1, 0, 1],
+        ...     "_EVENT_INDEX": [0, 1, 2, 0, 1, 2, 3, 4],
         ... })
-        >>> aggregate_event_bound_window(df, ToEventWindowBounds(True, "is_C", True, None))
+        >>> aggregate_event_bound_window(df, ToEventWindowBounds(True, "is_C", True, None)).drop("timestamp")
```
Comment (mmcdermott) [inline src/aces/aggregate.py:320]: I would keep the "timestamp" column in these tests, and just set the `pl.Config` params to print wider dataframes.

File: src/aces/aggregate.py [hunk 2]
```diff
@@ -305,71 +315,74 @@ def aggregate_event_bound_window(
...
+        │ 1          ┆ 1989-12-01 12:03:00 ┆ 1989-12-03 13:14:00 ┆ 1    ┆ 0    ┆ 0    ┆ 0            │
+        │ 1          ┆ null                ┆ null                ┆ 0    ┆ 0    ┆ 0    ┆ null         │
+        │ 1          ┆ null                ┆ null                ┆ 0    ┆ 0    ┆ 0    ┆ null         │
+        │ 2          ┆ 1989-12-02 12:03:00 ┆ 1989-12-06 15:17:00 ┆ 2    ┆ 1    ┆ 0    ┆ 1            │
+        │ 2          ┆ 1989-12-04 13:14:00 ┆ 1989-12-06 15:17:00 ┆ 1    ┆ 0    ┆ 0    ┆ 1            │
+        │ 2          ┆ 1989-12-06 15:17:00 ┆ 1989-12-10 03:07:00 ┆ 1    ┆ 2    ┆ 1    ┆ 3            │
+        │ 2          ┆ 1989-12-08 16:22:00 ┆ 1989-12-10 03:07:00 ┆ 0    ┆ 1    ┆ 0    ┆ 3            │
+        │ 2          ┆ null                ┆ null                ┆ 0    ┆ 0    ┆ 0    ┆ null         │
+        └────────────┴─────────────────────┴─────────────────────┴──────┴──────┴──────┴──────────────┘
+        >>> # unsure about event index for windows where start==end and no event (st_inclusive=False), null/#
```
Comment (mmcdermott) [inline src/aces/aggregate.py:352]: In those cases I think the window should be invalid and timestamps and event index should be null, no?

**Comment by mmcdermott** [comment]
I think more generally here, each aggregate function should return:
  1. The timestamp at the start of the window. If the window does not exist, this should be null.
  2. The timestamp at the end of the window. If the window does not exist, this should be null.
  3. The index of the first event inside the window. If there are no events in the window, this should be null.
  4. The index of the last event inside the window. If there are no events in the window, this should be null.

Note that the distinction between "the window existing" and "events existing" is as follows:
  1. A window is an interval in the patient timeline where (a) the left endpoint is <= the right endpoint and (b) either left or right endpoint can be open or closed. A window does not exist if either of these constraints are not satisfied. We can also modify this to say that "windows don't exist if they are provably empty -- e.g., `[12:01, 12:01)` of `(3:23, 3:23)`" or that "windows with equal left and right endpoints where both are open don't exist" as desired.
  2. Events are in the window if they occur with a timestamp within the interval defined by the window (respecting the openness or closedness of its bounds).

File: src/aces/aggregate.py (hunk 1 above)
Comment (justin13601) [reply src/aces/aggregate.py:320]: Actually I did this more to satistfy the pre-commit rules of lines not extending past 120 characters - should doctests still follow these rules? What's the best way to go around this?

**Comment by mmcdermott** [comment]
@justin13601, as we'll discuss more shortly, I'd actually suggest de-prioritizing this in favor of other things, and closing the PR for now as a result (with plans to re-visit later). Does that sound reasonable to you? If not, feel free to re-open.

**Pull request closed without merging**